File: sdks/javascript/tests/capabilities.test.ts
import { Capability, CapabilitySet } from '../src/capabilities';

describe('Capability', () => {
    test('should have unique values', () => {
        const values = Object.values(Capability).filter(v => typeof v === 'number') as number[];
        const uniqueValues = new Set(values);
        expect(values.length).toBe(uniqueValues.size);
    });

    test('should have NETWORK_OUTBOUND capability', () => {
        expect(Capability.NETWORK_OUTBOUND).toBeDefined();
        expect(typeof Capability.NETWORK_OUTBOUND).toBe('number');
    });

    test('should have NETWORK_INBOUND capability', () => {
        expect(Capability.NETWORK_INBOUND).toBeDefined();
        expect(typeof Capability.NETWORK_INBOUND).toBe('number');
    });

    test('should have STATE_READ capability', () => {
        expect(Capability.STATE_READ).toBeDefined();
        expect(typeof Capability.STATE_READ).toBe('number');
    });

    test('should have STATE_WRITE capability', () => {
        expect(Capability.STATE_WRITE).toBeDefined();
        expect(typeof Capability.STATE_WRITE).toBe('number');
    });

    test('should have FS_READ capability', () => {
        expect(Capability.FS_READ).toBeDefined();
        expect(typeof Capability.FS_READ).toBe('number');
    });

    test('should have FS_WRITE capability', () => {
        expect(Capability.FS_WRITE).toBeDefined();
        expect(typeof Capability.FS_WRITE).toBe('number');
    });

    test('should have ACTOR_MESSAGING capability', () => {
        expect(Capability.ACTOR_MESSAGING).toBeDefined();
        expect(typeof Capability.ACTOR_MESSAGING).toBe('number');
    });

    test('should have expected capability count', () => {
        const capabilities = Object.keys(Capability).filter(k => isNaN(Number(k)));
        expect(capabilities.length).toBe(13);
    });
});

describe('CapabilitySet', () => {
    test('should initialize empty', () => {
        const caps = new CapabilitySet();
        expect(caps.has(Capability.NETWORK_OUTBOUND)).toBe(false);
        expect(caps.has(Capability.STATE_READ)).toBe(false);
    });

    test('should initialize with capabilities', () => {
        const caps = new CapabilitySet(Capability.NETWORK_OUTBOUND, Capability.STATE_READ);
        expect(caps.has(Capability.NETWORK_OUTBOUND)).toBe(true);
        expect(caps.has(Capability.STATE_READ)).toBe(true);
        expect(caps.has(Capability.STATE_WRITE)).toBe(false);
    });

    test('should add capability', () => {
        const caps = new CapabilitySet();
        caps.add(Capability.NETWORK_OUTBOUND);
        expect(caps.has(Capability.NETWORK_OUTBOUND)).toBe(true);
    });

    test('should add multiple capabilities', () => {
        const caps = new CapabilitySet();
        caps.add(Capability.NETWORK_OUTBOUND);
        caps.add(Capability.STATE_READ);
        caps.add(Capability.STATE_WRITE);

        expect(caps.has(Capability.NETWORK_OUTBOUND)).toBe(true);
        expect(caps.has(Capability.STATE_READ)).toBe(true);
        expect(caps.has(Capability.STATE_WRITE)).toBe(true);
        expect(caps.has(Capability.FS_READ)).toBe(false);
    });

    test('should detect network capability', () => {
        const capsWithNetwork = new CapabilitySet(Capability.NETWORK_OUTBOUND);
        expect(capsWithNetwork.hasNetwork()).toBe(true);

        const capsWithInbound = new CapabilitySet(Capability.NETWORK_INBOUND);
        expect(capsWithInbound.hasNetwork()).toBe(true);

        const capsWithoutNetwork = new CapabilitySet(Capability.STATE_READ);
        expect(capsWithoutNetwork.hasNetwork()).toBe(false);
    });

    test('should detect state capability', () => {
        const capsWithState = new CapabilitySet(Capability.STATE_READ);
        expect(capsWithState.hasState()).toBe(true);

        const capsWithWrite = new CapabilitySet(Capability.STATE_WRITE);
        expect(capsWithWrite.hasState()).toBe(true);

        const capsWithoutState = new CapabilitySet(Capability.NETWORK_OUTBOUND);
        expect(capsWithoutState.hasState()).toBe(false);
    });

    test('should return false for missing capability', () => {
        const caps = new CapabilitySet(Capability.NETWORK_OUTBOUND);
        expect(caps.has(Capability.STATE_READ)).toBe(false);
    });

    test('should not duplicate when adding same capability', () => {
        const caps = new CapabilitySet();
        caps.add(Capability.NETWORK_OUTBOUND);
        caps.add(Capability.NETWORK_OUTBOUND);
        expect(caps.has(Capability.NETWORK_OUTBOUND)).toBe(true);
    });

    test('should add all capabilities', () => {
        const caps = new CapabilitySet();
        Object.values(Capability).forEach(cap => {
            if (typeof cap === 'number') {
                caps.add(cap);
            }
        });

        Object.keys(Capability).forEach(key => {
            if (isNaN(Number(key))) {
                expect(caps.has((Capability as any)[key as any])).toBe(true);
            }
        });
    });
});

File: sdks/go/aether/capabilities_test.go
package aether

import "testing"

func TestCapability_String(t *testing.T) {
	tests := []struct {
		cap      Capability
		expected string
	}{
		{CapabilityNetworkOutbound, "NETWORK_OUTBOUND"},
		{CapabilityNetworkInbound, "NETWORK_INBOUND"},
		{CapabilityStateRead, "STATE_READ"},
		{CapabilityStateWrite, "STATE_WRITE"},
		{CapabilityFSRead, "FS_READ"},
		{CapabilityFSWrite, "FS_WRITE"},
		{CapabilityActorMessaging, "ACTOR_MESSAGING"},
		{CapabilityLog, "LOG"},
		{CapabilityTime, "TIME"},
		{CapabilityRandom, "RANDOM"},
		{CapabilityEnvironment, "ENVIRONMENT"},
		{CapabilityHTTPClient, "HTTP_CLIENT"},
		{CapabilityHTTPServer, "HTTP_SERVER"},
		{CapabilityProcessSpawn, "PROCESS_SPAWN"},
		{Capability(999), "UNKNOWN"},
		{Capability(-1), "UNKNOWN"},
	}
	for _, tt := range tests {
		if got := tt.cap.String(); got != tt.expected {
			t.Errorf("Capability(%d).String() = %q, want %q", tt.cap, got, tt.expected)
		}
	}
}

func TestNewCapabilitySet(t *testing.T) {
	tests := []struct {
		name         string
		caps         []Capability
		checkHas     Capability
		checkHasBool bool
	}{
		{"empty", nil, CapabilityLog, false},
		{"single", []Capability{CapabilityLog}, CapabilityLog, true},
		{"multiple", []Capability{CapabilityNetworkOutbound, CapabilityStateRead}, CapabilityNetworkOutbound, true},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			cs := NewCapabilitySet(tt.caps...)
			if cs.Has(tt.checkHas) != tt.checkHasBool {
				t.Errorf("Has(%v) = %v, want %v", tt.checkHas, cs.Has(tt.checkHas), tt.checkHasBool)
			}
		})
	}
}

func TestCapabilitySet_Add(t *testing.T) {
	cs := NewCapabilitySet()
	cs.Add(CapabilityLog)
	if !cs.Has(CapabilityLog) {
		t.Error("expected LOG capability after Add")
	}
	cs.Add(CapabilityLog)
	if !cs.Has(CapabilityLog) {
		t.Error("re-adding should keep capability")
	}
}

func TestCapabilitySet_Has(t *testing.T) {
	cs := NewCapabilitySet(CapabilityNetworkOutbound, CapabilityLog)

	if !cs.Has(CapabilityNetworkOutbound) {
		t.Error("expected NETWORK_OUTBOUND")
	}
	if !cs.Has(CapabilityLog) {
		t.Error("expected LOG")
	}
	if cs.Has(CapabilityFSRead) {
		t.Error("did not expect FS_READ")
	}
}

func TestCapabilitySet_HasNetwork(t *testing.T) {
	tests := []struct {
		name string
		caps []Capability
		want bool
	}{
		{"outbound only", []Capability{CapabilityNetworkOutbound}, true},
		{"inbound only", []Capability{CapabilityNetworkInbound}, true},
		{"both", []Capability{CapabilityNetworkOutbound, CapabilityNetworkInbound}, true},
		{"neither", []Capability{CapabilityLog}, false},
		{"empty", nil, false},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			cs := NewCapabilitySet(tt.caps...)
			if got := cs.HasNetwork(); got != tt.want {
				t.Errorf("HasNetwork() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestCapabilitySet_HasState(t *testing.T) {
	tests := []struct {
		name string
		caps []Capability
		want bool
	}{
		{"read only", []Capability{CapabilityStateRead}, true},
		{"write only", []Capability{CapabilityStateWrite}, true},
		{"both", []Capability{CapabilityStateRead, CapabilityStateWrite}, true},
		{"neither", []Capability{CapabilityLog}, false},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			cs := NewCapabilitySet(tt.caps...)
			if got := cs.HasState(); got != tt.want {
				t.Errorf("HasState() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestCapabilitySet_HasFS(t *testing.T) {
	tests := []struct {
		name string
		caps []Capability
		want bool
	}{
		{"read only", []Capability{CapabilityFSRead}, true},
		{"write only", []Capability{CapabilityFSWrite}, true},
		{"neither", []Capability{CapabilityLog}, false},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			cs := NewCapabilitySet(tt.caps...)
			if got := cs.HasFS(); got != tt.want {
				t.Errorf("HasFS() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestCapabilitySet_HasHTTP(t *testing.T) {
	tests := []struct {
		name string
		caps []Capability
		want bool
	}{
		{"client only", []Capability{CapabilityHTTPClient}, true},
		{"server only", []Capability{CapabilityHTTPServer}, true},
		{"neither", []Capability{CapabilityLog}, false},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			cs := NewCapabilitySet(tt.caps...)
			if got := cs.HasHTTP(); got != tt.want {
				t.Errorf("HasHTTP() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestCapabilitySet_All(t *testing.T) {
	caps := []Capability{CapabilityLog, CapabilityTime, CapabilityRandom}
	cs := NewCapabilitySet(caps...)

	all := cs.All()
	if len(all) != len(caps) {
		t.Errorf("expected %d capabilities, got %d", len(caps), len(all))
	}

	found := make(map[Capability]bool)
	for _, c := range all {
		found[c] = true
	}
	for _, c := range caps {
		if !found[c] {
			t.Errorf("expected capability %v in All()", c)
		}
	}
}

func TestEmptyCapabilitySet(t *testing.T) {
	cs := EmptyCapabilitySet()
	if cs.Has(CapabilityLog) {
		t.Error("empty set should not have any capabilities")
	}
	if cs.HasNetwork() {
		t.Error("empty set should not have network")
	}
	all := cs.All()
	if len(all) != 0 {
		t.Errorf("expected empty All(), got %d", len(all))
	}
}

func TestAllCapabilities(t *testing.T) {
	cs := AllCapabilities()
	expectedCount := 14
	all := cs.All()
	if len(all) != expectedCount {
		t.Errorf("expected %d capabilities, got %d", expectedCount, len(all))
	}
	if !cs.Has(CapabilityNetworkOutbound) {
		t.Error("expected NETWORK_OUTBOUND in AllCapabilities")
	}
	if !cs.Has(CapabilityProcessSpawn) {
		t.Error("expected PROCESS_SPAWN in AllCapabilities")
	}
}

func TestCapabilitySet_All_HasDuplicates(t *testing.T) {
	cs := NewCapabilitySet(CapabilityLog, CapabilityLog)
	all := cs.All()
	if len(all) != 1 {
		t.Errorf("expected 1 unique capability, got %d", len(all))
	}
}

func TestCapabilitySet_NilMap(t *testing.T) {
	cs := &CapabilitySet{caps: make(map[Capability]bool)}
	if cs.Has(CapabilityLog) {
		t.Error("empty map should not have capabilities")
	}
}
